Support for Tuya Dimmers (Tuya TX / RX)

<!-- READ THIS FIRST:
 - This is for feature requests only, for issues please go to the issues repository.
 - Please be as descriptive as possible, especially use-cases that can otherwise not be solved boost the problem's priority.

 DO NOT DELETE ANY TEXT from this template! Otherwise the issue may be closed without a comment.
-->

**Describe the problem you have/What new integration you would like**
<!-- A description of what you want to happen. -->
Implementation of the Tuya Dimmers functionality in ESPHome that is present in Tasmota firmware.

**Please describe your use case for this integration and alternatives you've tried:**
<!-- A clear and concise description of what the problem is. -->
I have some Tuya dimmers at home that seem to use a secondary micro controller connected to the GPIO pins of the ESP. Tasmota implements this, but it would be great if they could be migrated to ESPHome.

**Additional context**
<!-- Add any other context about the feature request here. -->
I have found some of the resources from the Tasmota docs that look to be applicable:

https://github.com/arendst/Sonoff-Tasmota/wiki/Tuya-Dimmer

https://github.com/arendst/Sonoff-Tasmota/blob/development/sonoff/xdrv_16_tuyadimmer.ino


<!-- LEAVE THIS LINE AS-IS AND DON'T DELETE IT, OTHERWISE THE FEATURE REQUEST WILL BE CLOSED AUTOMATICALLY. -->

@placidorevilla When I add this line to tuyadimmer_lightoutput.h,
```
      if (checksum != byte) {
        ESP_LOGW(TAG, "Checksum doesn't match: 0x%02X!=0x%02X", byte, checksum);
        this->dump_packet_();
        return false;
      }
    }
```
 compilation fails.
`
Compiling /data/erc310/.pioenvs/erc310/src/main.cpp.o
In file included from src/main.cpp:13:0:
src/tuyadimmer_lightoutput.h: In member function 'esphome::optional<bool> TuyaDimmerLightOutput::check_byte_() const':
src/tuyadimmer_lightoutput.h:182:28: error: passing 'const TuyaDimmerLightOutput' as 'this' argument of 'void TuyaDimmerLightOutput::dump_packet_()' discards qualifiers [-fpermissive]
this->dump_packet_();
^
*** [/data/erc310/.pioenvs/erc310/src/main.cpp.o] Error 1
========================== [ERROR] Took 3.27 seconds ==========================
`

 What am I doing wrong?

@ssieb error compilation if I set min_value 20

`light:
  - platform: "tuya"
    name: "dim1"
    dimmer: 3
    switch: 1
    min_value: 20
    max_value: 255
    gamma_correct: 1.0
    default_transition_length: 0s`

`Linking .pioenvs/erc310/firmware.elf
.pioenvs/erc310/src/main.cpp.o:(.text.setup+0x68): undefined reference to `esphome::tuya::TuyaLight::set_min_value(unsigned int)'
.pioenvs/erc310/src/main.cpp.o:(.text.setup+0x6c): undefined reference to `esphome::tuya::TuyaLight::set_max_value(unsigned int)'
.pioenvs/erc310/src/main.cpp.o: In function `setup':
main.cpp:(.text.setup+0x3cb): undefined reference to `esphome::tuya::TuyaLight::set_min_value(unsigned int)'
main.cpp:(.text.setup+0x3d6): undefined reference to `esphome::tuya::TuyaLight::set_max_value(unsigned int)'
collect2: error: ld returned 1 exit status
*** [.pioenvs/erc310/firmware.elf] Error 1
============================= [FAILED] Took 5.74 seconds =============================`

--- a/esphome/components/esp32_ble_tracker/esp32_ble_tracker.cpp
+++ b/esphome/components/esp32_ble_tracker/esp32_ble_tracker.cpp
@@ -208,19 +208,7 @@ void ESP32BLETracker::gap_scan_result(const esp_ble_gap_cb_param_t::ble_scan_res
 }
 
 std::string hexencode(const std::string &raw_data) {
-  char buf[20];
-  std::string res;
-  for (size_t i = 0; i < raw_data.size(); i++) {
-    if (i + 1 != raw_data.size()) {
-      sprintf(buf, "0x%02X.", static_cast<uint8_t>(raw_data[i]));
-    } else {
-      sprintf(buf, "0x%02X ", static_cast<uint8_t>(raw_data[i]));
-    }
-    res += buf;
-  }
-  sprintf(buf, "(%zu)", raw_data.size());
-  res += buf;
-  return res;
+  return hexencode(reinterpret_cast<const uint8_t *>(raw_data.c_str()), raw_data.size());
 }
 
 ESPBTUUID::ESPBTUUID() : uuid_() {}

--- a/esphome/components/esp32_ble_tracker/esp32_ble_tracker.cpp
+++ b/esphome/components/esp32_ble_tracker/esp32_ble_tracker.cpp
@@ -1,6 +1,7 @@
 #include "esp32_ble_tracker.h"
 #include "esphome/core/log.h"
 #include "esphome/core/application.h"
+#include "esphome/core/helpers.h"
 
 #ifdef ARDUINO_ARCH_ESP32
 
@@ -207,7 +208,7 @@ void ESP32BLETracker::gap_scan_result(const esp_ble_gap_cb_param_t::ble_scan_res
   }
 }
 
-std::string hexencode(const std::string &raw_data) {
+std::string hexencode_string(const std::string &raw_data) {
   return hexencode(reinterpret_cast<const uint8_t *>(raw_data.c_str()), raw_data.size());
 }
 
@@ -320,15 +321,15 @@ void ESPBTDevice::parse_scan_rst(const esp_ble_gap_cb_param_t::ble_scan_result_e
   for (auto uuid : this->service_uuids_) {
     ESP_LOGVV(TAG, "  Service UUID: %s", uuid.to_string().c_str());
   }
-  ESP_LOGVV(TAG, "  Manufacturer data: %s", hexencode(this->manufacturer_data_).c_str());
-  ESP_LOGVV(TAG, "  Service data: %s", hexencode(this->service_data_).c_str());
+  ESP_LOGVV(TAG, "  Manufacturer data: %s", hexencode_string(this->manufacturer_data_).c_str());
+  ESP_LOGVV(TAG, "  Service data: %s", hexencode_string(this->service_data_).c_str());
 
   if (this->service_data_uuid_.has_value()) {
     ESP_LOGVV(TAG, "  Service Data UUID: %s", this->service_data_uuid_->to_string().c_str());
   }
 
   ESP_LOGVV(TAG, "Adv data: %s",
-            hexencode(std::string(reinterpret_cast<const char *>(param.ble_adv), param.adv_data_len)).c_str());
+            hexencode_string(std::string(reinterpret_cast<const char *>(param.ble_adv), param.adv_data_len)).c_str());
 #endif
 }
 void ESPBTDevice::parse_adv_(const esp_ble_gap_cb_param_t::ble_scan_result_evt_param &param) {

--- a/esphome/components/tuya/__init__.py
+++ b/esphome/components/tuya/__init__.py
@@ -1,0 +1,20 @@
+import esphome.codegen as cg
+import esphome.config_validation as cv
+from esphome.components import uart
+from esphome.const import CONF_ID
+
+DEPENDENCIES = ['uart']
+
+tuya_ns = cg.esphome_ns.namespace('tuya')
+TUYA = tuya_ns.class_('Tuya', cg.Component, uart.UARTDevice)
+
+CONF_TUYA_ID = 'tuya_id'
+CONFIG_SCHEMA = cv.Schema({
+    cv.GenerateID(): cv.declare_id(TUYA),
+}).extend(cv.COMPONENT_SCHEMA).extend(uart.UART_DEVICE_SCHEMA)
+
+
+def to_code(config):
+    var = cg.new_Pvariable(config[CONF_ID])
+    yield cg.register_component(var, config)
+    yield uart.register_uart_device(var, config)

--- a/esphome/components/tuya/__init__.py
+++ b/esphome/components/tuya/__init__.py
@@ -6,11 +6,11 @@ from esphome.const import CONF_ID
 DEPENDENCIES = ['uart']
 
 tuya_ns = cg.esphome_ns.namespace('tuya')
-TUYA = tuya_ns.class_('Tuya', cg.Component, uart.UARTDevice)
+Tuya = tuya_ns.class_('Tuya', cg.Component, uart.UARTDevice)
 
 CONF_TUYA_ID = 'tuya_id'
 CONFIG_SCHEMA = cv.Schema({
-    cv.GenerateID(): cv.declare_id(TUYA),
+    cv.GenerateID(): cv.declare_id(Tuya),
 }).extend(cv.COMPONENT_SCHEMA).extend(uart.UART_DEVICE_SCHEMA)
 
 

--- a/esphome/components/tuya/light/__init__.py
+++ b/esphome/components/tuya/light/__init__.py
@@ -1,0 +1,38 @@
+from esphome.components import light
+import esphome.config_validation as cv
+import esphome.codegen as cg
+from esphome.const import CONF_OUTPUT_ID, CONF_MIN_VALUE, CONF_MAX_VALUE
+from .. import tuya_ns, CONF_TUYA_ID, TUYA
+
+DEPENDENCIES = ['tuya']
+
+CONF_SWITCH = "switch"
+CONF_DIMMER = "dimmer"
+
+TUYALIGHT = tuya_ns.class_('TuyaLight', light.LightOutput, cg.Component)
+
+CONFIG_SCHEMA = light.BRIGHTNESS_ONLY_LIGHT_SCHEMA.extend({
+    cv.GenerateID(CONF_OUTPUT_ID): cv.declare_id(TUYALIGHT),
+    cv.GenerateID(CONF_TUYA_ID): cv.use_id(TUYA),
+    cv.Required(CONF_DIMMER): cv.int_,
+    cv.Optional(CONF_SWITCH): cv.int_,
+    cv.Optional(CONF_MIN_VALUE): cv.int_,
+    cv.Optional(CONF_MAX_VALUE): cv.int_,
+}).extend(cv.COMPONENT_SCHEMA)
+
+
+def to_code(config):
+    var = cg.new_Pvariable(config[CONF_OUTPUT_ID])
+    yield cg.register_component(var, config)
+    yield light.register_light(var, config)
+
+    if CONF_DIMMER in config:
+        cg.add(var.set_dimmer(config[CONF_DIMMER]))
+    if CONF_SWITCH in config:
+        cg.add(var.set_switch(config[CONF_SWITCH]))
+    if CONF_MIN_VALUE in config:
+        cg.add(var.set_min_value(config[CONF_MIN_VALUE]))
+    if CONF_MAX_VALUE in config:
+        cg.add(var.set_max_value(config[CONF_MAX_VALUE]))
+    paren = yield cg.get_variable(config[CONF_TUYA_ID])
+    cg.add(var.set_tuya_parent(paren))

--- a/esphome/components/tuya/light/__init__.py
+++ b/esphome/components/tuya/light/__init__.py
@@ -9,10 +9,10 @@ DEPENDENCIES = ['tuya']
 CONF_DIMMER_DATAPOINT = "dimmer_datapoint"
 CONF_SWITCH_DATAPOINT = "switch_datapoint"
 
-TUYALIGHT = tuya_ns.class_('TuyaLight', light.LightOutput, cg.Component)
+TuyaLight = tuya_ns.class_('TuyaLight', light.LightOutput, cg.Component)
 
 CONFIG_SCHEMA = light.BRIGHTNESS_ONLY_LIGHT_SCHEMA.extend({
-    cv.GenerateID(CONF_OUTPUT_ID): cv.declare_id(TUYALIGHT),
+    cv.GenerateID(CONF_OUTPUT_ID): cv.declare_id(TuyaLight),
     cv.GenerateID(CONF_TUYA_ID): cv.use_id(TUYA),
     cv.Required(CONF_DIMMER_DATAPOINT): cv.uint8_t,
     cv.Optional(CONF_SWITCH_DATAPOINT): cv.uint8_t,
@@ -27,9 +27,9 @@ def to_code(config):
     yield light.register_light(var, config)
 
     if CONF_DIMMER_DATAPOINT in config:
-        cg.add(var.set_dimmer(config[CONF_DIMMER_DATAPOINT]))
+        cg.add(var.set_dimmer_id(config[CONF_DIMMER_DATAPOINT]))
     if CONF_SWITCH_DATAPOINT in config:
-        cg.add(var.set_switch(config[CONF_SWITCH_DATAPOINT]))
+        cg.add(var.set_switch_id(config[CONF_SWITCH_DATAPOINT]))
     if CONF_MIN_VALUE in config:
         cg.add(var.set_min_value(config[CONF_MIN_VALUE]))
     if CONF_MAX_VALUE in config:

--- a/esphome/components/tuya/light/__init__.py
+++ b/esphome/components/tuya/light/__init__.py
@@ -6,16 +6,16 @@ from .. import tuya_ns, CONF_TUYA_ID, TUYA
 
 DEPENDENCIES = ['tuya']
 
-CONF_SWITCH = "switch"
-CONF_DIMMER = "dimmer"
+CONF_DIMMER_DATAPOINT = "dimmer_datapoint"
+CONF_SWITCH_DATAPOINT = "switch_datapoint"
 
 TUYALIGHT = tuya_ns.class_('TuyaLight', light.LightOutput, cg.Component)
 
 CONFIG_SCHEMA = light.BRIGHTNESS_ONLY_LIGHT_SCHEMA.extend({
     cv.GenerateID(CONF_OUTPUT_ID): cv.declare_id(TUYALIGHT),
     cv.GenerateID(CONF_TUYA_ID): cv.use_id(TUYA),
-    cv.Required(CONF_DIMMER): cv.int_,
-    cv.Optional(CONF_SWITCH): cv.int_,
+    cv.Required(CONF_DIMMER_DATAPOINT): cv.uint8_t,
+    cv.Optional(CONF_SWITCH_DATAPOINT): cv.uint8_t,
     cv.Optional(CONF_MIN_VALUE): cv.int_,
     cv.Optional(CONF_MAX_VALUE): cv.int_,
 }).extend(cv.COMPONENT_SCHEMA)
@@ -26,10 +26,10 @@ def to_code(config):
     yield cg.register_component(var, config)
     yield light.register_light(var, config)
 
-    if CONF_DIMMER in config:
-        cg.add(var.set_dimmer(config[CONF_DIMMER]))
-    if CONF_SWITCH in config:
-        cg.add(var.set_switch(config[CONF_SWITCH]))
+    if CONF_DIMMER_DATAPOINT in config:
+        cg.add(var.set_dimmer(config[CONF_DIMMER_DATAPOINT]))
+    if CONF_SWITCH_DATAPOINT in config:
+        cg.add(var.set_switch(config[CONF_SWITCH_DATAPOINT]))
     if CONF_MIN_VALUE in config:
         cg.add(var.set_min_value(config[CONF_MIN_VALUE]))
     if CONF_MAX_VALUE in config:

--- a/esphome/components/tuya/light/__init__.py
+++ b/esphome/components/tuya/light/__init__.py
@@ -2,7 +2,7 @@ from esphome.components import light
 import esphome.config_validation as cv
 import esphome.codegen as cg
 from esphome.const import CONF_OUTPUT_ID, CONF_MIN_VALUE, CONF_MAX_VALUE
-from .. import tuya_ns, CONF_TUYA_ID, TUYA
+from .. import tuya_ns, CONF_TUYA_ID, Tuya
 
 DEPENDENCIES = ['tuya']
 
@@ -13,7 +13,7 @@ TuyaLight = tuya_ns.class_('TuyaLight', light.LightOutput, cg.Component)
 
 CONFIG_SCHEMA = light.BRIGHTNESS_ONLY_LIGHT_SCHEMA.extend({
     cv.GenerateID(CONF_OUTPUT_ID): cv.declare_id(TuyaLight),
-    cv.GenerateID(CONF_TUYA_ID): cv.use_id(TUYA),
+    cv.GenerateID(CONF_TUYA_ID): cv.use_id(Tuya),
     cv.Required(CONF_DIMMER_DATAPOINT): cv.uint8_t,
     cv.Optional(CONF_SWITCH_DATAPOINT): cv.uint8_t,
     cv.Optional(CONF_MIN_VALUE): cv.int_,

--- a/esphome/components/tuya/light/tuya_light.cpp
+++ b/esphome/components/tuya/light/tuya_light.cpp
@@ -7,87 +7,77 @@ namespace tuya {
 static const char *TAG = "tuya.light";
 
 void TuyaLight::setup() {
-  if (parent_ == nullptr)
-    return;
-  if (dimmer_id_ != -1) {
-    parent_->register_listener(dimmer_id_, this);
-    dimmer_value_ = parent_->get_dp_value(dimmer_id_);
+  if (this->dimmer_id_.has_value()) {
+    this->parent_->register_listener(*this->dimmer_id_, [this](TuyaDatapoint datapoint) {
+      auto call = this->state_->make_call();
+      call.set_brightness(float(datapoint.value_uint) / this->max_value_);
+      call.perform();
+    });
   }
-  if (switch_id_ != -1) {
-    parent_->register_listener(switch_id_, this);
-    switch_value_ = parent_->get_dp_value(switch_id_);
+  if (switch_id_.has_value()) {
+    this->parent_->register_listener(*this->switch_id_, [this](TuyaDatapoint datapoint) {
+      auto call = this->state_->make_call();
+      call.set_state(datapoint.value_bool);
+      call.perform();
+    });
   }
 }
 
 void TuyaLight::dump_config() {
   ESP_LOGCONFIG(TAG, "Tuya Dimmer:");
-  if (dimmer_id_ != -1)
-    ESP_LOGCONFIG(TAG, "   dimmer is dpid %d", dimmer_id_);
-  if (switch_id_ != -1)
-    ESP_LOGCONFIG(TAG, "   switch is dpid %d", switch_id_);
-}
-
-void TuyaLight::set_dimmer(int dpid) { dimmer_id_ = dpid; }
-
-void TuyaLight::set_switch(int dpid) { switch_id_ = dpid; }
-
-void TuyaLight::set_tuya_parent(Tuya *parent) { parent_ = parent; }
-
-void TuyaLight::set_min_value(uint32_t value) { min_value_ = value; }
-
-void TuyaLight::set_max_value(uint32_t value) { max_value_ = value; }
-
-void TuyaLight::dp_update(int dpid, uint32_t value) {
-  ESP_LOGV(TAG, "dp_update(%d, %d)", dpid, value);
-  if (dpid == dimmer_id_)
-    update_dimmer_(value);
-  if (dpid == switch_id_)
-    update_switch_(value);
-}
-
-void TuyaLight::update_dimmer_(uint32_t value) {
-  ESP_LOGV(TAG, "dimmer update to %d", value);
-  state_->make_call().set_brightness(float(value) / max_value_).perform();
-}
-
-void TuyaLight::update_switch_(uint32_t value) {
-  ESP_LOGV(TAG, "switch update to %d", value);
-  if (value > 0)
-    state_->turn_on().perform();
-  else
-    state_->turn_off().perform();
+  if (this->dimmer_id_.has_value())
+    ESP_LOGCONFIG(TAG, "   Dimmer has datapoint ID %u", *this->dimmer_id_);
+  if (this->switch_id_.has_value())
+    ESP_LOGCONFIG(TAG, "   Switch has datapoint ID %u", *this->switch_id_);
 }
 
 light::LightTraits TuyaLight::get_traits() {
   auto traits = light::LightTraits();
-  traits.set_supports_brightness(true);
+  traits.set_supports_brightness(this->dimmer_id_.has_value());
   return traits;
 }
 
 void TuyaLight::setup_state(light::LightState *state) { state_ = state; }
 
 void TuyaLight::write_state(light::LightState *state) {
-  auto val = state->current_values;
-  uint32_t value = (uint32_t)(val.get_brightness() * max_value_);
-  if (!val.is_on()) {
-    if (switch_id_ != -1) {
-      parent_->set_dp_value(switch_id_, 0);
-      switch_value_ = 0;
-    } else if (dimmer_id_ != -1) {
-      parent_->set_dp_value(dimmer_id_, 0);
-      dimmer_value_ = 0;
+  float brightness;
+  state->current_values_as_brightness(&brightness);
+
+  if (brightness == 0.0f) {
+    // turning off, first try via switch (if exists), then dimmer
+    if (switch_id_.has_value()) {
+      TuyaDatapoint datapoint{};
+      datapoint.id = *this->switch_id_;
+      datapoint.type = TuyaDatapointType::BOOLEAN;
+      datapoint.value_bool = false;
+
+      parent_->set_datapoint_value(datapoint);
+    } else if (dimmer_id_.has_value()) {
+      TuyaDatapoint datapoint{};
+      datapoint.id = *this->dimmer_id_;
+      datapoint.type = TuyaDatapointType::INTEGER;
+      datapoint.value_int = 0;
+      parent_->set_datapoint_value(datapoint);
     }
     return;
   }
-  if (value < min_value_)
-    value = min_value_;
-  if (dimmer_id_ != -1) {
-    parent_->set_dp_value(dimmer_id_, value);
-    dimmer_value_ = value;
+
+  auto brightness_int = static_cast<uint32_t>(brightness * this->max_value_);
+  brightness_int = std::min(brightness_int, this->min_value_);
+
+  if (this->dimmer_id_.has_value()) {
+    TuyaDatapoint datapoint{};
+    datapoint.id = *this->dimmer_id_;
+    datapoint.type = TuyaDatapointType::INTEGER;
+    datapoint.value_int = brightness_int;
+    parent_->set_datapoint_value(datapoint);
   }
-  if ((switch_id_ != -1) && (switch_value_ == 0)) {
-    parent_->set_dp_value(switch_id_, 1);
-    switch_value_ = 1;
+  if (this->switch_id_ == 0) {
+    TuyaDatapoint datapoint{};
+    datapoint.id = *this->switch_id_;
+    datapoint.type = TuyaDatapointType::BOOLEAN;
+    datapoint.value_bool = true;
+    parent_->set_datapoint_value(datapoint);
   }
 }
 

--- a/esphome/components/tuya/light/tuya_light.cpp
+++ b/esphome/components/tuya/light/tuya_light.cpp
@@ -1,0 +1,95 @@
+#include "esphome/core/log.h"
+#include "tuya_light.h"
+
+namespace esphome {
+namespace tuya {
+
+static const char *TAG = "tuya.light";
+
+void TuyaLight::setup() {
+  if (parent_ == nullptr)
+    return;
+  if (dimmer_id_ != -1) {
+    parent_->register_listener(dimmer_id_, this);
+    dimmer_value_ = parent_->get_dp_value(dimmer_id_);
+  }
+  if (switch_id_ != -1) {
+    parent_->register_listener(switch_id_, this);
+    switch_value_ = parent_->get_dp_value(switch_id_);
+  }
+}
+
+void TuyaLight::dump_config() {
+  ESP_LOGCONFIG(TAG, "Tuya Dimmer:");
+  if (dimmer_id_ != -1)
+    ESP_LOGCONFIG(TAG, "   dimmer is dpid %d", dimmer_id_);
+  if (switch_id_ != -1)
+    ESP_LOGCONFIG(TAG, "   switch is dpid %d", switch_id_);
+}
+
+void TuyaLight::set_dimmer(int dpid) { dimmer_id_ = dpid; }
+
+void TuyaLight::set_switch(int dpid) { switch_id_ = dpid; }
+
+void TuyaLight::set_tuya_parent(Tuya *parent) { parent_ = parent; }
+
+void TuyaLight::set_min_value(uint32_t value) { min_value_ = value; }
+
+void TuyaLight::set_max_value(uint32_t value) { max_value_ = value; }
+
+void TuyaLight::dp_update(int dpid, uint32_t value) {
+  ESP_LOGV(TAG, "dp_update(%d, %d)", dpid, value);
+  if (dpid == dimmer_id_)
+    update_dimmer_(value);
+  if (dpid == switch_id_)
+    update_switch_(value);
+}
+
+void TuyaLight::update_dimmer_(uint32_t value) {
+  ESP_LOGV(TAG, "dimmer update to %d", value);
+  state_->make_call().set_brightness(float(value) / max_value_).perform();
+}
+
+void TuyaLight::update_switch_(uint32_t value) {
+  ESP_LOGV(TAG, "switch update to %d", value);
+  if (value > 0)
+    state_->turn_on().perform();
+  else
+    state_->turn_off().perform();
+}
+
+light::LightTraits TuyaLight::get_traits() {
+  auto traits = light::LightTraits();
+  traits.set_supports_brightness(true);
+  return traits;
+}
+
+void TuyaLight::setup_state(light::LightState *state) { state_ = state; }
+
+void TuyaLight::write_state(light::LightState *state) {
+  auto val = state->current_values;
+  uint32_t value = (uint32_t)(val.get_brightness() * max_value_);
+  if (!val.is_on()) {
+    if (switch_id_ != -1) {
+      parent_->set_dp_value(switch_id_, 0);
+      switch_value_ = 0;
+    } else if (dimmer_id_ != -1) {
+      parent_->set_dp_value(dimmer_id_, 0);
+      dimmer_value_ = 0;
+    }
+    return;
+  }
+  if (value < min_value_)
+    value = min_value_;
+  if (dimmer_id_ != -1) {
+    parent_->set_dp_value(dimmer_id_, value);
+    dimmer_value_ = value;
+  }
+  if ((switch_id_ != -1) && (switch_value_ == 0)) {
+    parent_->set_dp_value(switch_id_, 1);
+    switch_value_ = 1;
+  }
+}
+
+}  // namespace tuya
+}  // namespace esphome

--- a/esphome/components/tuya/light/tuya_light.cpp
+++ b/esphome/components/tuya/light/tuya_light.cpp
@@ -63,7 +63,7 @@ void TuyaLight::write_state(light::LightState *state) {
   }
 
   auto brightness_int = static_cast<uint32_t>(brightness * this->max_value_);
-  brightness_int = std::min(brightness_int, this->min_value_);
+  brightness_int = std::max(brightness_int, this->min_value_);
 
   if (this->dimmer_id_.has_value()) {
     TuyaDatapoint datapoint{};
@@ -72,7 +72,7 @@ void TuyaLight::write_state(light::LightState *state) {
     datapoint.value_int = brightness_int;
     parent_->set_datapoint_value(datapoint);
   }
-  if (this->switch_id_ == 0) {
+  if (this->switch_id_.has_value()) {
     TuyaDatapoint datapoint{};
     datapoint.id = *this->switch_id_;
     datapoint.type = TuyaDatapointType::BOOLEAN;

--- a/esphome/components/tuya/light/tuya_light.h
+++ b/esphome/components/tuya/light/tuya_light.h
@@ -7,29 +7,26 @@
 namespace esphome {
 namespace tuya {
 
-class TuyaLight : public Component, public light::LightOutput, public TuyaListener {
+class TuyaLight : public Component, public light::LightOutput {
  public:
   void setup() override;
   void dump_config() override;
-  void set_dimmer(int dpid);
-  void set_switch(int dpid);
-  void set_tuya_parent(Tuya *parent);
-  void set_min_value(uint32_t value);
-  void set_max_value(uint32_t value);
+  void set_dimmer_id(uint8_t dimmer_id) { this->dimmer_id_ = dimmer_id; }
+  void set_switch_id(uint8_t switch_id) { this->switch_id_ = switch_id; }
+  void set_tuya_parent(Tuya *parent) { this->parent_ = parent; }
+  void set_min_value(uint32_t min_value) { min_value_ = min_value; }
+  void set_max_value(uint32_t max_value) { max_value_ = max_value; }
   light::LightTraits get_traits() override;
   void setup_state(light::LightState *state) override;
   void write_state(light::LightState *state) override;
-  void dp_update(int dpid, uint32_t value) override;
 
  protected:
   void update_dimmer_(uint32_t value);
   void update_switch_(uint32_t value);
 
   Tuya *parent_;
-  int dimmer_id_ = -1;
-  int switch_id_ = -1;
-  uint32_t dimmer_value_ = 0;
-  uint32_t switch_value_ = 0;
+  optional<uint8_t> dimmer_id_{};
+  optional<uint8_t> switch_id_{};
   uint32_t min_value_ = 0;
   uint32_t max_value_ = 255;
   light::LightState *state_{nullptr};

--- a/esphome/components/tuya/light/tuya_light.h
+++ b/esphome/components/tuya/light/tuya_light.h
@@ -1,0 +1,39 @@
+#pragma once
+
+#include "esphome/core/component.h"
+#include "esphome/components/tuya/tuya.h"
+#include "esphome/components/light/light_output.h"
+
+namespace esphome {
+namespace tuya {
+
+class TuyaLight : public Component, public light::LightOutput, public TuyaListener {
+ public:
+  void setup() override;
+  void dump_config() override;
+  void set_dimmer(int dpid);
+  void set_switch(int dpid);
+  void set_tuya_parent(Tuya *parent);
+  void set_min_value(uint32_t value);
+  void set_max_value(uint32_t value);
+  light::LightTraits get_traits() override;
+  void setup_state(light::LightState *state) override;
+  void write_state(light::LightState *state) override;
+  void dp_update(int dpid, uint32_t value) override;
+
+ protected:
+  void update_dimmer_(uint32_t value);
+  void update_switch_(uint32_t value);
+
+  Tuya *parent_;
+  int dimmer_id_ = -1;
+  int switch_id_ = -1;
+  uint32_t dimmer_value_ = 0;
+  uint32_t switch_value_ = 0;
+  uint32_t min_value_ = 0;
+  uint32_t max_value_ = 255;
+  light::LightState *state_{nullptr};
+};
+
+}  // namespace tuya
+}  // namespace esphome

--- a/esphome/components/tuya/tuya.cpp
+++ b/esphome/components/tuya/tuya.cpp
@@ -1,0 +1,261 @@
+#include "tuya.h"
+#include "esphome/core/log.h"
+
+namespace esphome {
+namespace tuya {
+
+static const uint8_t TUYA_CMD_HEARTBEAT = 0;
+static const uint8_t TUYA_CMD_QUERY_PRODUCT = 1;
+static const uint8_t TUYA_CMD_MCU_CONF = 2;
+static const uint8_t TUYA_CMD_WIFI_STATE = 3;
+static const uint8_t TUYA_CMD_WIFI_RESET = 4;
+static const uint8_t TUYA_CMD_WIFI_SELECT = 5;
+static const uint8_t TUYA_CMD_SET_DP = 6;
+static const uint8_t TUYA_CMD_STATE = 7;
+static const uint8_t TUYA_CMD_QUERY_STATE = 8;
+
+static const uint8_t TUYA_TYPE_BOOL = 1;
+static const uint8_t TUYA_TYPE_INT = 2;
+
+static const char *TAG = "tuya";
+
+/* Ask the MCU for a status update in order to find out
+ * what devices are available */
+void Tuya::setup() {
+  uint32_t last = millis();
+  uint8_t c;
+
+  this->send_command_(TUYA_CMD_MCU_CONF);
+  // there is no end of list indicator so wait until no more data is sent
+  while (millis() - last < 50) {
+    if (this->read_byte(&c)) {
+      last = millis();
+      this->handle_char_(c);
+    }
+  }
+  in_setup_ = false;
+  c = 3;
+  this->send_command_(TUYA_CMD_WIFI_STATE, 1, &c);  // set wifi state LED on
+}
+
+void Tuya::loop() {
+  uint32_t now = millis();
+  if (now - last_hb_ > 1000) {
+    last_hb_ = now;
+    this->send_command_(TUYA_CMD_HEARTBEAT);
+  }
+  while (this->available()) {
+    uint8_t c;
+    if (!this->read_byte(&c))
+      return;
+    ESP_LOGVV(TAG, "received char %02x", c);
+    this->handle_char_(c);
+  }
+}
+
+void Tuya::dump_config() {
+  ESP_LOGCONFIG(TAG, "Tuya:");
+  if ((gpio_status_ != -1) || (gpio_reset_ != -1))
+    ESP_LOGCONFIG(TAG, "  GPIO MCU configuration not supported!");
+  for (auto &info : dp_info_) {
+    if (info.type == TUYA_TYPE_BOOL)
+      ESP_LOGCONFIG(TAG, "  %d: switch", info.id);
+    else if (info.type == TUYA_TYPE_INT)
+      ESP_LOGCONFIG(TAG, "  %d: int value", info.id);
+    else
+      ESP_LOGCONFIG(TAG, "  %d: unknown", info.id);
+  }
+}
+
+void Tuya::register_listener(int dpid, TuyaListener *listener) {
+  for (auto &info : dp_info_)
+    if (info.id == dpid)
+      info.listener = listener;
+}
+
+char *buffer_to_string(int count, uint8_t *buffer) {
+  static char RET[80];
+  for (int i = 0; i < count; i++)
+    sprintf(RET + i * 3, "%02x ", buffer[i]);
+  return RET;
+}
+
+void Tuya::handle_char_(int c) {
+  static int STATE = 0;
+  static uint8_t BUFFER[20];
+  static int BUFPTR = 0;
+  static uint8_t CHECKSUM;
+  static uint8_t VERSION, COMMAND, COUNT;
+
+  switch (STATE) {
+    case 0:
+      if (c == 0x55)
+        STATE = 1;
+      break;
+    case 1:
+      if (c == 0xaa)
+        STATE = 2;
+      else
+        STATE = 0;
+      break;
+    case 2:
+      VERSION = c;
+      CHECKSUM = 0xff + c;  // 0x55 + 0xaa = 0xff
+      STATE = 3;
+      break;
+    case 3:
+      COMMAND = c;
+      CHECKSUM += c;
+      STATE = 4;
+      break;
+    case 4:
+      COUNT = c;
+      CHECKSUM += c;
+      STATE = 5;
+      break;
+    case 5:
+      COUNT = (COUNT << 8) + c;
+      if (COUNT > 20) {
+        ESP_LOGE(TAG, "received response with %d bytes of data", COUNT);
+        STATE = 0;
+        break;
+      }
+      CHECKSUM += c;
+      BUFPTR = 0;
+      if (COUNT > 0)
+        STATE = 6;
+      else
+        STATE = 7;
+      break;
+    case 6:
+      BUFFER[BUFPTR++] = c;
+      CHECKSUM += c;
+      if (BUFPTR == COUNT)
+        STATE = 7;
+      break;
+    case 7:
+      STATE = 0;
+      ESP_LOGV(TAG, "command %d version %d count %d data [%s] checksum %d %d", COMMAND, VERSION, COUNT,
+               buffer_to_string(COUNT, BUFFER), CHECKSUM, c);
+      if (CHECKSUM != c)
+        ESP_LOGE(TAG, "checksum error");
+      else
+        this->handle_command_(COMMAND, VERSION, COUNT, BUFFER);
+      break;
+  }
+}
+
+void Tuya::handle_command_(uint8_t command, uint8_t version, int count, uint8_t *buffer) {
+  int id;
+  int type;
+  uint32_t value;
+
+  switch (command) {
+    case TUYA_CMD_HEARTBEAT:
+      if (buffer[0] == 0) {
+        ESP_LOGI(TAG, "MCU restarted");
+        this->send_command_(TUYA_CMD_QUERY_STATE);
+      }
+      break;
+    case TUYA_CMD_QUERY_PRODUCT:
+      ESP_LOGI(TAG, "product code: %s", buffer);
+      break;
+    case TUYA_CMD_MCU_CONF:
+      if (count != 0) {
+        gpio_status_ = buffer[0];
+        gpio_reset_ = buffer[1];
+      }
+      this->send_command_(TUYA_CMD_QUERY_STATE);
+      break;
+    case TUYA_CMD_WIFI_STATE:
+      break;
+    case TUYA_CMD_WIFI_RESET:
+      ESP_LOGE(TAG, "TUYA_CMD_WIFI_RESET is not handled");
+      break;
+    case TUYA_CMD_WIFI_SELECT:
+      ESP_LOGE(TAG, "TUYA_CMD_WIFI_SELECT is not handled");
+      break;
+    case TUYA_CMD_SET_DP:
+      break;
+    case TUYA_CMD_STATE:
+      id = buffer[0];
+      type = buffer[1];
+      value = 0;
+      if (type == TUYA_TYPE_BOOL)
+        value = buffer[4];
+      else if (type == TUYA_TYPE_INT)
+        value = (buffer[4] << 24) + (buffer[5] << 16) + (buffer[6] << 8) + buffer[7];
+      if (in_setup_) {
+        dp_info_.push_back({id, type, value, nullptr});
+        break;
+      }
+      ESP_LOGV(TAG, "dp %d update to %d", id, value);
+      for (auto &info : dp_info_) {
+        if (info.id == id) {
+          if (info.value != value) {
+            info.value = value;
+            if (info.listener != nullptr)
+              info.listener->dp_update(id, value);
+          }
+        }
+      }
+      break;
+    case TUYA_CMD_QUERY_STATE:
+      break;
+    default:
+      ESP_LOGE(TAG, "invalid command (%02x) received", command);
+  }
+}
+
+void Tuya::send_command_(uint8_t command, int count, uint8_t *buffer) {
+  uint8_t checksum;
+
+  this->write_byte(0x55);
+  this->write_byte(0xaa);
+  this->write_byte(0x00);  // version 0
+  this->write_byte(command);
+  this->write_byte(0x00);
+  this->write_byte(count);
+  checksum = 0xff + command + count;  // 0x55 + 0xaa = 0xff
+  for (int i = 0; i < count; i++) {
+    this->write_byte(buffer[i]);
+    checksum += buffer[i];
+  }
+  this->write_byte(checksum);
+}
+
+void Tuya::set_dp_value(int dpid, uint32_t value) {
+  uint8_t buffer[8];
+  for (auto &info : dp_info_) {
+    if (info.id == dpid) {
+      int count = 4;
+      buffer[0] = dpid;
+      buffer[1] = info.type;
+      buffer[2] = 0;  // high size
+      if (info.type == TUYA_TYPE_BOOL) {
+        buffer[3] = 1;
+        buffer[4] = value;
+        count++;
+      } else if (info.type == TUYA_TYPE_INT) {
+        buffer[3] = 4;
+        buffer[4] = value >> 24;
+        buffer[5] = value >> 16;
+        buffer[6] = value >> 8;
+        buffer[7] = value;
+        count += 4;
+      }
+      send_command_(TUYA_CMD_SET_DP, count, buffer);
+      break;
+    }
+  }
+}
+
+uint32_t Tuya::get_dp_value(int dpid) {
+  for (auto &info : dp_info_)
+    if (info.id == dpid)
+      return info.value;
+  return 0;
+}
+
+}  // namespace tuya
+}  // namespace esphome

--- a/esphome/components/tuya/tuya.cpp
+++ b/esphome/components/tuya/tuya.cpp
@@ -1,54 +1,40 @@
 #include "tuya.h"
 #include "esphome/core/log.h"
+#include "esphome/core/helpers.h"
 
 namespace esphome {
 namespace tuya {
 
-static const uint8_t TUYA_CMD_HEARTBEAT = 0;
-static const uint8_t TUYA_CMD_QUERY_PRODUCT = 1;
-static const uint8_t TUYA_CMD_MCU_CONF = 2;
-static const uint8_t TUYA_CMD_WIFI_STATE = 3;
-static const uint8_t TUYA_CMD_WIFI_RESET = 4;
-static const uint8_t TUYA_CMD_WIFI_SELECT = 5;
-static const uint8_t TUYA_CMD_SET_DP = 6;
-static const uint8_t TUYA_CMD_STATE = 7;
-static const uint8_t TUYA_CMD_QUERY_STATE = 8;
-
-static const uint8_t TUYA_TYPE_BOOL = 1;
-static const uint8_t TUYA_TYPE_INT = 2;
-
 static const char *TAG = "tuya";
 
-/* Ask the MCU for a status update in order to find out
- * what devices are available */
 void Tuya::setup() {
-  uint32_t last = millis();
-  uint8_t c;
+  this->send_empty_command_(TuyaCommandType::MCU_CONF);
+  this->last_setup_timestamp_ = millis();
 
-  this->send_command_(TUYA_CMD_MCU_CONF);
-  // there is no end of list indicator so wait until no more data is sent
-  while (millis() - last < 50) {
-    if (this->read_byte(&c)) {
-      last = millis();
-      this->handle_char_(c);
-    }
-  }
-  in_setup_ = false;
-  c = 3;
-  this->send_command_(TUYA_CMD_WIFI_STATE, 1, &c);  // set wifi state LED on
+  this->set_interval("heartbeat", 1000, [this] { this->send_empty_command_(TuyaCommandType::HEARTBEAT); });
 }
 
 void Tuya::loop() {
-  uint32_t now = millis();
-  if (now - last_hb_ > 1000) {
-    last_hb_ = now;
-    this->send_command_(TUYA_CMD_HEARTBEAT);
+  const uint32_t now = millis();
+  if (this->in_setup_) {
+    if (now - this->last_setup_timestamp_ >= 50) {
+      this->in_setup_ = false;
+      uint8_t c = 0x03;
+      this->send_command_(TuyaCommandType::WIFI_STATE, &c, 1);  // set wifi state LED on
+    } else {
+      // still waiting for messages
+      uint8_t c;
+      while (this->read_byte(&c)) {
+        this->last_setup_timestamp_ = now;
+        this->handle_char_(c);
+      }
+      return;
+    }
   }
+
   while (this->available()) {
     uint8_t c;
-    if (!this->read_byte(&c))
-      return;
-    ESP_LOGVV(TAG, "received char %02x", c);
+    this->read_byte(&c);
     this->handle_char_(c);
   }
 }
@@ -57,204 +43,237 @@ void Tuya::dump_config() {
   ESP_LOGCONFIG(TAG, "Tuya:");
   if ((gpio_status_ != -1) || (gpio_reset_ != -1))
     ESP_LOGCONFIG(TAG, "  GPIO MCU configuration not supported!");
-  for (auto &info : dp_info_) {
-    if (info.type == TUYA_TYPE_BOOL)
-      ESP_LOGCONFIG(TAG, "  %d: switch", info.id);
-    else if (info.type == TUYA_TYPE_INT)
-      ESP_LOGCONFIG(TAG, "  %d: int value", info.id);
+  for (auto &info : this->datapoints_) {
+    if (info.type == TuyaDatapointType::BOOLEAN)
+      ESP_LOGCONFIG(TAG, "  %d: switch (value: %s)", info.id, ONOFF(info.value_bool));
+    else if (info.type == TuyaDatapointType::INTEGER)
+      ESP_LOGCONFIG(TAG, "  %d: int value (value: %d)", info.id, info.value_int);
     else
       ESP_LOGCONFIG(TAG, "  %d: unknown", info.id);
   }
 }
 
-void Tuya::register_listener(int dpid, TuyaListener *listener) {
-  for (auto &info : dp_info_)
-    if (info.id == dpid)
-      info.listener = listener;
+bool Tuya::validate_message_() {
+  uint32_t at = this->rx_message_.size() - 1;
+  auto *data = &this->rx_message_[0];
+  uint8_t new_byte = data[at];
+
+  // Byte 0: HEADER1 (always 0x55)
+  if (at == 0)
+    return new_byte == 0x55;
+  // Byte 1: HEADER2 (always 0xAA)
+  if (at == 1)
+    return new_byte == 0xAA;
+
+  // Byte 2: VERSION
+  // no validation for the following fields:
+  uint8_t version = data[2];
+  if (at == 2)
+    return true;
+  // Byte 3: COMMAND
+  uint8_t command = data[3];
+  if (at == 3)
+    return true;
+
+  // Byte 4: LENGTH1
+  // Byte 5: LENGTH2
+  if (at <= 5)
+    // no validation for these fields
+    return true;
+
+  uint16_t length = (uint16_t(data[4]) << 8) | (uint16_t(data[5]));
+
+  // wait until all data is read
+  if (at - 6 < length)
+    return true;
+
+  // Byte 6+LEN: CHECKSUM - sum of all bytes (including header) modulo 256
+  uint8_t rx_checksum = new_byte;
+  uint8_t calc_checksum = 0;
+  for (uint32_t i = 0; i < 6 + length; i++)
+    calc_checksum += data[i];
+
+  if (rx_checksum == calc_checksum) {
+    // valid message
+    const uint8_t *message_data = data + 6;
+    ESP_LOGV(TAG, "Received Tuya: CMD=0x%02X VERSION=%u DATA=[%s]", command, version,
+             hexencode(message_data, length).c_str());
+    this->handle_command_(command, version, message_data, length);
+
+    // return false to reset to HEADER1
+    return false;
+  }
+
+  return false;
 }
 
-char *buffer_to_string(int count, uint8_t *buffer) {
-  static char RET[80];
-  for (int i = 0; i < count; i++)
-    sprintf(RET + i * 3, "%02x ", buffer[i]);
-  return RET;
-}
-
-void Tuya::handle_char_(int c) {
-  static int STATE = 0;
-  static uint8_t BUFFER[20];
-  static int BUFPTR = 0;
-  static uint8_t CHECKSUM;
-  static uint8_t VERSION, COMMAND, COUNT;
-
-  switch (STATE) {
-    case 0:
-      if (c == 0x55)
-        STATE = 1;
-      break;
-    case 1:
-      if (c == 0xaa)
-        STATE = 2;
-      else
-        STATE = 0;
-      break;
-    case 2:
-      VERSION = c;
-      CHECKSUM = 0xff + c;  // 0x55 + 0xaa = 0xff
-      STATE = 3;
-      break;
-    case 3:
-      COMMAND = c;
-      CHECKSUM += c;
-      STATE = 4;
-      break;
-    case 4:
-      COUNT = c;
-      CHECKSUM += c;
-      STATE = 5;
-      break;
-    case 5:
-      COUNT = (COUNT << 8) + c;
-      if (COUNT > 20) {
-        ESP_LOGE(TAG, "received response with %d bytes of data", COUNT);
-        STATE = 0;
-        break;
-      }
-      CHECKSUM += c;
-      BUFPTR = 0;
-      if (COUNT > 0)
-        STATE = 6;
-      else
-        STATE = 7;
-      break;
-    case 6:
-      BUFFER[BUFPTR++] = c;
-      CHECKSUM += c;
-      if (BUFPTR == COUNT)
-        STATE = 7;
-      break;
-    case 7:
-      STATE = 0;
-      ESP_LOGV(TAG, "command %d version %d count %d data [%s] checksum %d %d", COMMAND, VERSION, COUNT,
-               buffer_to_string(COUNT, BUFFER), CHECKSUM, c);
-      if (CHECKSUM != c)
-        ESP_LOGE(TAG, "checksum error");
-      else
-        this->handle_command_(COMMAND, VERSION, COUNT, BUFFER);
-      break;
+void Tuya::handle_char_(uint8_t c) {
+  this->rx_message_.push_back(c);
+  if (!this->validate_message_()) {
+    this->rx_message_.clear();
   }
 }
 
-void Tuya::handle_command_(uint8_t command, uint8_t version, int count, uint8_t *buffer) {
-  int id;
-  int type;
-  uint32_t value;
-
-  switch (command) {
-    case TUYA_CMD_HEARTBEAT:
+void Tuya::handle_command_(uint8_t command, uint8_t version, const uint8_t *buffer, size_t len) {
+  switch ((TuyaCommandType) command) {
+    case TuyaCommandType::HEARTBEAT:
+      ESP_LOGV(TAG, "MCU Heartbeat (0x%02X)", buffer[0]);
       if (buffer[0] == 0) {
         ESP_LOGI(TAG, "MCU restarted");
-        this->send_command_(TUYA_CMD_QUERY_STATE);
+        this->send_empty_command_(TuyaCommandType::QUERY_STATE);
       }
       break;
-    case TUYA_CMD_QUERY_PRODUCT:
-      ESP_LOGI(TAG, "product code: %s", buffer);
+    case TuyaCommandType::QUERY_PRODUCT: {
+      // check it is a valid string
+      bool valid = false;
+      for (int i = 0; i < len; i++) {
+        if (buffer[i] == 0x00) {
+          valid = true;
+          break;
+        }
+      }
+      if (valid) {
+        ESP_LOGD(TAG, "Tuya Product Code: %s", reinterpret_cast<const char *>(buffer));
+      }
       break;
-    case TUYA_CMD_MCU_CONF:
-      if (count != 0) {
+    }
+    case TuyaCommandType::MCU_CONF:
+      if (len >= 2) {
         gpio_status_ = buffer[0];
         gpio_reset_ = buffer[1];
       }
-      this->send_command_(TUYA_CMD_QUERY_STATE);
+      this->send_empty_command_(TuyaCommandType::QUERY_STATE);
       break;
-    case TUYA_CMD_WIFI_STATE:
+    case TuyaCommandType::WIFI_STATE:
       break;
-    case TUYA_CMD_WIFI_RESET:
+    case TuyaCommandType::WIFI_RESET:
       ESP_LOGE(TAG, "TUYA_CMD_WIFI_RESET is not handled");
       break;
-    case TUYA_CMD_WIFI_SELECT:
+    case TuyaCommandType::WIFI_SELECT:
       ESP_LOGE(TAG, "TUYA_CMD_WIFI_SELECT is not handled");
       break;
-    case TUYA_CMD_SET_DP:
+    case TuyaCommandType::SET_DATAPOINT:
       break;
-    case TUYA_CMD_STATE:
-      id = buffer[0];
-      type = buffer[1];
-      value = 0;
-      if (type == TUYA_TYPE_BOOL)
-        value = buffer[4];
-      else if (type == TUYA_TYPE_INT)
-        value = (buffer[4] << 24) + (buffer[5] << 16) + (buffer[6] << 8) + buffer[7];
-      if (in_setup_) {
-        dp_info_.push_back({id, type, value, nullptr});
-        break;
-      }
-      ESP_LOGV(TAG, "dp %d update to %d", id, value);
-      for (auto &info : dp_info_) {
-        if (info.id == id) {
-          if (info.value != value) {
-            info.value = value;
-            if (info.listener != nullptr)
-              info.listener->dp_update(id, value);
-          }
-        }
-      }
+    case TuyaCommandType::STATE: {
+      this->handle_datapoint_(buffer, len);
+
       break;
-    case TUYA_CMD_QUERY_STATE:
+    }
+    case TuyaCommandType::QUERY_STATE:
       break;
     default:
       ESP_LOGE(TAG, "invalid command (%02x) received", command);
   }
 }
 
-void Tuya::send_command_(uint8_t command, int count, uint8_t *buffer) {
-  uint8_t checksum;
+void Tuya::handle_datapoint_(const uint8_t *buffer, size_t len) {
+  if (len < 2)
+    return;
 
-  this->write_byte(0x55);
-  this->write_byte(0xaa);
-  this->write_byte(0x00);  // version 0
-  this->write_byte(command);
-  this->write_byte(0x00);
-  this->write_byte(count);
-  checksum = 0xff + command + count;  // 0x55 + 0xaa = 0xff
-  for (int i = 0; i < count; i++) {
-    this->write_byte(buffer[i]);
-    checksum += buffer[i];
+  TuyaDatapoint datapoint{};
+  datapoint.id = buffer[0];
+  datapoint.type = (TuyaDatapointType) buffer[1];
+
+  const uint8_t *data = buffer + 2;
+  size_t data_len = len - 2;
+
+  switch (datapoint.type) {
+    case TuyaDatapointType::BOOLEAN:
+      if (data_len != 1)
+        return;
+      datapoint.value_bool = data[0];
+      break;
+    case TuyaDatapointType::INTEGER:
+      if (data_len != 4)
+        return;
+      datapoint.value_uint =
+          (uint32_t(data[0]) << 24) | (uint32_t(data[1]) << 16) | (uint32_t(data[2]) << 8) | (uint32_t(data[3]) << 0);
+      break;
+    case TuyaDatapointType::ENUM:
+      if (data_len != 1)
+        return;
+      datapoint.value_enum = data[0];
+      break;
+    case TuyaDatapointType::BITMASK:
+      if (data_len != 4)
+        return;
+      datapoint.value_bitmask = (uint16_t(data[0]) << 8) | (uint16_t(data[1]) << 0);
+      break;
+    default:
+      return;
   }
+  ESP_LOGV(TAG, "Datapoint %u update to %u", datapoint.id, datapoint.value_uint);
+
+  // Update internal datapoints
+  for (auto &other : this->datapoints_)
+    if (other.id == datapoint.id)
+      other = datapoint;
+
+  // Run through listeners
+  for (auto &listener : this->listeners_)
+    if (listener.datapoint_id == datapoint.id)
+      listener.on_datapoint(datapoint);
+}
+
+void Tuya::send_command_(TuyaCommandType command, const uint8_t *buffer, uint16_t len) {
+  uint8_t len_hi = len >> 8;
+  uint8_t len_lo = len >> 0;
+  this->write_array({0x55, 0xAA,
+                     0x00,  // version
+                     (uint8_t) command, len_hi, len_lo});
+  if (len != 0)
+    this->write_array(buffer, len);
+
+  uint8_t checksum = 0x55 + 0xAA + (uint8_t) command + len_hi + len_lo;
+  for (int i = 0; i < len; i++)
+    checksum += buffer[i];
   this->write_byte(checksum);
 }
 
-void Tuya::set_dp_value(int dpid, uint32_t value) {
-  uint8_t buffer[8];
-  for (auto &info : dp_info_) {
-    if (info.id == dpid) {
-      int count = 4;
-      buffer[0] = dpid;
-      buffer[1] = info.type;
-      buffer[2] = 0;  // high size
-      if (info.type == TUYA_TYPE_BOOL) {
-        buffer[3] = 1;
-        buffer[4] = value;
-        count++;
-      } else if (info.type == TUYA_TYPE_INT) {
-        buffer[3] = 4;
-        buffer[4] = value >> 24;
-        buffer[5] = value >> 16;
-        buffer[6] = value >> 8;
-        buffer[7] = value;
-        count += 4;
-      }
-      send_command_(TUYA_CMD_SET_DP, count, buffer);
+void Tuya::set_datapoint_value(TuyaDatapoint datapoint) {
+  std::vector<uint8_t> buffer;
+  buffer.push_back(datapoint.id);
+  buffer.push_back(static_cast<uint8_t>(datapoint.type));
+
+  std::vector<uint8_t> data;
+  switch (datapoint.type) {
+    case TuyaDatapointType::BOOLEAN:
+      data.push_back(datapoint.value_bool);
       break;
-    }
+    case TuyaDatapointType::INTEGER:
+      data.push_back(datapoint.value_uint >> 24);
+      data.push_back(datapoint.value_uint >> 16);
+      data.push_back(datapoint.value_uint >> 8);
+      data.push_back(datapoint.value_uint >> 0);
+      break;
+    case TuyaDatapointType::ENUM:
+      data.push_back(datapoint.value_enum);
+      break;
+    case TuyaDatapointType::BITMASK:
+      data.push_back(datapoint.value_bitmask >> 8);
+      data.push_back(datapoint.value_bitmask >> 0);
+      break;
+    default:
+      return;
   }
+
+  buffer.push_back(data.size() >> 8);
+  buffer.push_back(data.size() >> 0);
+  buffer.insert(buffer.end(), data.begin(), data.end());
+  this->send_command_(TuyaCommandType::SET_DATAPOINT, buffer.data(), buffer.size());
 }
 
-uint32_t Tuya::get_dp_value(int dpid) {
-  for (auto &info : dp_info_)
-    if (info.id == dpid)
-      return info.value;
-  return 0;
+void Tuya::register_listener(uint8_t datapoint_id, const std::function<void(TuyaDatapoint)> &func) {
+  auto listener = TuyaDatapointListener{
+      .datapoint_id = datapoint_id,
+      .on_datapoint = func,
+  };
+  this->listeners_.push_back(listener);
+
+  // Run through existing datapoints
+  for (auto &datapoint : this->datapoints_)
+    if (datapoint.id == datapoint_id)
+      func(datapoint);
 }
 
 }  // namespace tuya

--- a/esphome/components/tuya/tuya.cpp
+++ b/esphome/components/tuya/tuya.cpp
@@ -93,17 +93,19 @@ bool Tuya::validate_message_() {
   for (uint32_t i = 0; i < 6 + length; i++)
     calc_checksum += data[i];
 
-  if (rx_checksum == calc_checksum) {
-    // valid message
-    const uint8_t *message_data = data + 6;
-    ESP_LOGV(TAG, "Received Tuya: CMD=0x%02X VERSION=%u DATA=[%s]", command, version,
-             hexencode(message_data, length).c_str());
-    this->handle_command_(command, version, message_data, length);
-
-    // return false to reset to HEADER1
+  if (rx_checksum != calc_checksum) {
+    ESP_LOGW(TAG, "Tuya Received invalid message checksum %02X!=%02X",
+             rx_checksum, calc_checksum);
     return false;
   }
 
+  // valid message
+  const uint8_t *message_data = data + 6;
+  ESP_LOGV(TAG, "Received Tuya: CMD=0x%02X VERSION=%u DATA=[%s]", command, version,
+           hexencode(message_data, length).c_str());
+  this->handle_command_(command, version, message_data, length);
+
+  // return false to reset rx buffer
   return false;
 }
 
@@ -195,7 +197,7 @@ void Tuya::handle_datapoint_(const uint8_t *buffer, size_t len) {
       datapoint.value_enum = data[0];
       break;
     case TuyaDatapointType::BITMASK:
-      if (data_len != 4)
+      if (data_len != 2)
         return;
       datapoint.value_bitmask = (uint16_t(data[0]) << 8) | (uint16_t(data[1]) << 0);
       break;
@@ -205,9 +207,20 @@ void Tuya::handle_datapoint_(const uint8_t *buffer, size_t len) {
   ESP_LOGV(TAG, "Datapoint %u update to %u", datapoint.id, datapoint.value_uint);
 
   // Update internal datapoints
-  for (auto &other : this->datapoints_)
-    if (other.id == datapoint.id)
+  bool found = false;
+  for (auto &other : this->datapoints_) {
+    if (other.id == datapoint.id) {
       other = datapoint;
+      found = true;
+    }
+  }
+  if (!found) {
+    this->datapoints_.push_back(datapoint);
+    // New datapoint found, reprint dump_config after a delay.
+    this->set_timeout("datapoint_dump", 100, [this]{
+      this->dump_config();
+    });
+  }
 
   // Run through listeners
   for (auto &listener : this->listeners_)

--- a/esphome/components/tuya/tuya.cpp
+++ b/esphome/components/tuya/tuya.cpp
@@ -45,11 +45,11 @@ void Tuya::dump_config() {
     ESP_LOGCONFIG(TAG, "  GPIO MCU configuration not supported!");
   for (auto &info : this->datapoints_) {
     if (info.type == TuyaDatapointType::BOOLEAN)
-      ESP_LOGCONFIG(TAG, "  %d: switch (value: %s)", info.id, ONOFF(info.value_bool));
+      ESP_LOGCONFIG(TAG, "  Datapoint %d: switch (value: %s)", info.id, ONOFF(info.value_bool));
     else if (info.type == TuyaDatapointType::INTEGER)
-      ESP_LOGCONFIG(TAG, "  %d: int value (value: %d)", info.id, info.value_int);
+      ESP_LOGCONFIG(TAG, "  Datapoint %d: int value (value: %d)", info.id, info.value_int);
     else
-      ESP_LOGCONFIG(TAG, "  %d: unknown", info.id);
+      ESP_LOGCONFIG(TAG, "  Datapoint %d: unknown", info.id);
   }
 }
 

--- a/esphome/components/tuya/tuya.cpp
+++ b/esphome/components/tuya/tuya.cpp
@@ -9,29 +9,10 @@ static const char *TAG = "tuya";
 
 void Tuya::setup() {
   this->send_empty_command_(TuyaCommandType::MCU_CONF);
-  this->last_setup_timestamp_ = millis();
-
   this->set_interval("heartbeat", 1000, [this] { this->send_empty_command_(TuyaCommandType::HEARTBEAT); });
 }
 
 void Tuya::loop() {
-  const uint32_t now = millis();
-  if (this->in_setup_) {
-    if (now - this->last_setup_timestamp_ >= 50) {
-      this->in_setup_ = false;
-      uint8_t c = 0x03;
-      this->send_command_(TuyaCommandType::WIFI_STATE, &c, 1);  // set wifi state LED on
-    } else {
-      // still waiting for messages
-      uint8_t c;
-      while (this->read_byte(&c)) {
-        this->last_setup_timestamp_ = now;
-        this->handle_char_(c);
-      }
-      return;
-    }
-  }
-
   while (this->available()) {
     uint8_t c;
     this->read_byte(&c);
@@ -48,6 +29,10 @@ void Tuya::dump_config() {
       ESP_LOGCONFIG(TAG, "  Datapoint %d: switch (value: %s)", info.id, ONOFF(info.value_bool));
     else if (info.type == TuyaDatapointType::INTEGER)
       ESP_LOGCONFIG(TAG, "  Datapoint %d: int value (value: %d)", info.id, info.value_int);
+    else if (info.type == TuyaDatapointType::ENUM)
+      ESP_LOGCONFIG(TAG, "  Datapoint %d: enum (value: %d)", info.id, info.value_enum);
+    else if (info.type == TuyaDatapointType::BITMASK)
+      ESP_LOGCONFIG(TAG, "  Datapoint %d: bitmask (value: %x)", info.id, info.value_bitmask);
     else
       ESP_LOGCONFIG(TAG, "  Datapoint %d: unknown", info.id);
   }
@@ -116,6 +101,7 @@ void Tuya::handle_char_(uint8_t c) {
 }
 
 void Tuya::handle_command_(uint8_t command, uint8_t version, const uint8_t *buffer, size_t len) {
+  uint8_t c;
   switch ((TuyaCommandType) command) {
     case TuyaCommandType::HEARTBEAT:
       ESP_LOGV(TAG, "MCU Heartbeat (0x%02X)", buffer[0]);
@@ -143,6 +129,10 @@ void Tuya::handle_command_(uint8_t command, uint8_t version, const uint8_t *buff
         gpio_status_ = buffer[0];
         gpio_reset_ = buffer[1];
       }
+      // set wifi state LED to off or on depending on the MCU firmware
+      // but it shouldn't be blinking
+      c = 0x3;
+      this->send_command_(TuyaCommandType::WIFI_STATE, &c, 1);
       this->send_empty_command_(TuyaCommandType::QUERY_STATE);
       break;
     case TuyaCommandType::WIFI_STATE:
@@ -157,7 +147,6 @@ void Tuya::handle_command_(uint8_t command, uint8_t version, const uint8_t *buff
       break;
     case TuyaCommandType::STATE: {
       this->handle_datapoint_(buffer, len);
-
       break;
     }
     case TuyaCommandType::QUERY_STATE:
@@ -174,9 +163,15 @@ void Tuya::handle_datapoint_(const uint8_t *buffer, size_t len) {
   TuyaDatapoint datapoint{};
   datapoint.id = buffer[0];
   datapoint.type = (TuyaDatapointType) buffer[1];
+  datapoint.value_uint = 0;
 
-  const uint8_t *data = buffer + 2;
-  size_t data_len = len - 2;
+  size_t data_size = (buffer[2] << 8) + buffer[3];
+  const uint8_t *data = buffer + 4;
+  size_t data_len = len - 4;
+  if (data_size != data_len) {
+    ESP_LOGW(TAG, "invalid datapoint update");
+    return;
+  }
 
   switch (datapoint.type) {
     case TuyaDatapointType::BOOLEAN:
@@ -243,6 +238,14 @@ void Tuya::send_command_(TuyaCommandType command, const uint8_t *buffer, uint16_
 void Tuya::set_datapoint_value(TuyaDatapoint datapoint) {
   std::vector<uint8_t> buffer;
   ESP_LOGV(TAG, "Datapoint %u set to %u", datapoint.id, datapoint.value_uint);
+  for (auto &other : this->datapoints_) {
+    if (other.id == datapoint.id) {
+      if (other.value_uint == datapoint.value_uint) {
+        ESP_LOGV(TAG, "Not sending unchanged value");
+        return;
+      }
+    }
+  }
   buffer.push_back(datapoint.id);
   buffer.push_back(static_cast<uint8_t>(datapoint.type));
 

--- a/esphome/components/tuya/tuya.cpp
+++ b/esphome/components/tuya/tuya.cpp
@@ -242,6 +242,7 @@ void Tuya::send_command_(TuyaCommandType command, const uint8_t *buffer, uint16_
 
 void Tuya::set_datapoint_value(TuyaDatapoint datapoint) {
   std::vector<uint8_t> buffer;
+  ESP_LOGV(TAG, "Datapoint %u set to %u", datapoint.id, datapoint.value_uint);
   buffer.push_back(datapoint.id);
   buffer.push_back(static_cast<uint8_t>(datapoint.type));
 

--- a/esphome/components/tuya/tuya.cpp
+++ b/esphome/components/tuya/tuya.cpp
@@ -94,8 +94,7 @@ bool Tuya::validate_message_() {
     calc_checksum += data[i];
 
   if (rx_checksum != calc_checksum) {
-    ESP_LOGW(TAG, "Tuya Received invalid message checksum %02X!=%02X",
-             rx_checksum, calc_checksum);
+    ESP_LOGW(TAG, "Tuya Received invalid message checksum %02X!=%02X", rx_checksum, calc_checksum);
     return false;
   }
 
@@ -217,9 +216,7 @@ void Tuya::handle_datapoint_(const uint8_t *buffer, size_t len) {
   if (!found) {
     this->datapoints_.push_back(datapoint);
     // New datapoint found, reprint dump_config after a delay.
-    this->set_timeout("datapoint_dump", 100, [this]{
-      this->dump_config();
-    });
+    this->set_timeout("datapoint_dump", 100, [this] { this->dump_config(); });
   }
 
   // Run through listeners

--- a/esphome/components/tuya/tuya.h
+++ b/esphome/components/tuya/tuya.h
@@ -6,16 +6,42 @@
 namespace esphome {
 namespace tuya {
 
-class TuyaListener {
- public:
-  virtual void dp_update(int dpid, uint32_t value) = 0;
+enum class TuyaDatapointType : uint8_t {
+  RAW = 0x00,      // variable length
+  BOOLEAN = 0x01,  // 1 byte (0/1)
+  INTEGER = 0x02,  // 4 byte
+  STRING = 0x03,   // variable length
+  ENUM = 0x04,     // 1 byte
+  BITMASK = 0x05,  // 2 bytes
 };
 
-struct DpT {
-  int id;
-  int type;
-  uint32_t value;
-  TuyaListener *listener;
+struct TuyaDatapoint {
+  uint8_t id;
+  TuyaDatapointType type;
+  union {
+    bool value_bool;
+    int value_int;
+    uint32_t value_uint;
+    uint8_t value_enum;
+    uint16_t value_bitmask;
+  };
+};
+
+struct TuyaDatapointListener {
+  uint8_t datapoint_id;
+  std::function<void(TuyaDatapoint)> on_datapoint;
+};
+
+enum class TuyaCommandType : uint8_t {
+  HEARTBEAT = 0x00,
+  QUERY_PRODUCT = 0x01,
+  MCU_CONF = 0x02,
+  WIFI_STATE = 0x03,
+  WIFI_RESET = 0x04,
+  WIFI_SELECT = 0x05,
+  SET_DATAPOINT = 0x06,
+  STATE = 0x07,
+  QUERY_STATE = 0x08,
 };
 
 class Tuya : public Component, public uart::UARTDevice {
@@ -24,20 +50,26 @@ class Tuya : public Component, public uart::UARTDevice {
   void setup() override;
   void loop() override;
   void dump_config() override;
-  void register_listener(int dpid, TuyaListener *listener);
-  void set_dp_value(int dpid, uint32_t value);
-  uint32_t get_dp_value(int dpid);
+  void register_listener(uint8_t datapoint_id, const std::function<void(TuyaDatapoint)> &func);
+  void set_datapoint_value(TuyaDatapoint datapoint);
 
  protected:
-  void handle_char_(int c);
-  void handle_command_(uint8_t command, uint8_t version, int count, uint8_t *buffer);
-  void send_command_(uint8_t command, int count = 0, uint8_t *buffer = nullptr);
+  void handle_char_(uint8_t c);
+  void handle_datapoint_(const uint8_t *buffer, size_t len);
+  bool validate_message_();
+
+  void handle_command_(uint8_t command, uint8_t version, const uint8_t *buffer, size_t len);
+  void send_command_(TuyaCommandType command, const uint8_t *buffer, uint16_t len);
+  void send_empty_command_(TuyaCommandType command) { this->send_command_(command, nullptr, 0); }
 
   int gpio_status_ = -1;
   int gpio_reset_ = -1;
-  std::vector<DpT> dp_info_;
+  std::vector<TuyaDatapointListener> listeners_;
+  std::vector<TuyaDatapoint> datapoints_;
   bool in_setup_ = true;
-  uint32_t last_hb_ = 0;
+  uint32_t last_setup_timestamp_;
+
+  std::vector<uint8_t> rx_message_;
 };
 
 }  // namespace tuya

--- a/esphome/components/tuya/tuya.h
+++ b/esphome/components/tuya/tuya.h
@@ -1,0 +1,44 @@
+#pragma once
+
+#include "esphome/core/component.h"
+#include "esphome/components/uart/uart.h"
+
+namespace esphome {
+namespace tuya {
+
+class TuyaListener {
+ public:
+  virtual void dp_update(int dpid, uint32_t value) = 0;
+};
+
+struct DpT {
+  int id;
+  int type;
+  uint32_t value;
+  TuyaListener *listener;
+};
+
+class Tuya : public Component, public uart::UARTDevice {
+ public:
+  float get_setup_priority() const override { return setup_priority::HARDWARE; }
+  void setup() override;
+  void loop() override;
+  void dump_config() override;
+  void register_listener(int dpid, TuyaListener *listener);
+  void set_dp_value(int dpid, uint32_t value);
+  uint32_t get_dp_value(int dpid);
+
+ protected:
+  void handle_char_(int c);
+  void handle_command_(uint8_t command, uint8_t version, int count, uint8_t *buffer);
+  void send_command_(uint8_t command, int count = 0, uint8_t *buffer = nullptr);
+
+  int gpio_status_ = -1;
+  int gpio_reset_ = -1;
+  std::vector<DpT> dp_info_;
+  bool in_setup_ = true;
+  uint32_t last_hb_ = 0;
+};
+
+}  // namespace tuya
+}  // namespace esphome

--- a/esphome/components/tuya/tuya.h
+++ b/esphome/components/tuya/tuya.h
@@ -66,9 +66,6 @@ class Tuya : public Component, public uart::UARTDevice {
   int gpio_reset_ = -1;
   std::vector<TuyaDatapointListener> listeners_;
   std::vector<TuyaDatapoint> datapoints_;
-  bool in_setup_ = true;
-  uint32_t last_setup_timestamp_;
-
   std::vector<uint8_t> rx_message_;
 };
 

--- a/esphome/core/helpers.cpp
+++ b/esphome/core/helpers.cpp
@@ -314,4 +314,20 @@ std::array<uint8_t, 2> decode_uint16(uint16_t value) {
   return {msb, lsb};
 }
 
+std::string hexencode(const uint8_t *data, uint32_t len) {
+  char buf[20];
+  std::string res;
+  for (size_t i = 0; i < len; i++) {
+    if (i + 1 != len) {
+      sprintf(buf, "%02X.", data[i]);
+    } else {
+      sprintf(buf, "%02X ", data[i]);
+    }
+    res += buf;
+  }
+  sprintf(buf, "(%u)", len);
+  res += buf;
+  return res;
+}
+
 }  // namespace esphome

--- a/esphome/core/helpers.h
+++ b/esphome/core/helpers.h
@@ -156,6 +156,9 @@ enum ParseOnOffState {
 
 ParseOnOffState parse_on_off(const char *str, const char *on = nullptr, const char *off = nullptr);
 
+// Encode raw data to a human-readable string (for debugging)
+std::string hexencode(const uint8_t *data, uint32_t len);
+
 // https://stackoverflow.com/questions/7858817/unpacking-a-tuple-to-call-a-matching-function-pointer/7858971#7858971
 template<int...> struct seq {};                                       // NOLINT
 template<int N, int... S> struct gens : gens<N - 1, N - 1, S...> {};  // NOLINT

